language/move-core/types/src/language_storage.rs

```diff
@@ -79,6 +103,145 @@ impl<'a> From<&'a TypeTag> for SafeTypeTag<'a> {
     }
 }
 
+struct TypeTagVisitor(u8);
+
+use serde::de::{
+    self, DeserializeSeed, Deserializer, EnumAccess, MapAccess, SeqAccess, VariantAccess, Visitor,
+};
+
+impl<'de> Visitor<'de> for TypeTagVisitor {
+    type Value = TypeTag;
+
+    fn expecting(&self, formatter: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        formatter.write_str("`bool`, `u8`, `u64`, `u128`, `address`, `signer`, `vector`, `struct`")
+    }
+
+    fn visit_enum<A>(self, data: A) -> Result<Self::Value, A::Error>
+    where
+        A: EnumAccess<'de>,
+    {
+        let (variant_type, variant) = data.variant::<DeTypeTag>()?;
+        match variant_type {
+            DeTypeTag::Bool => {
+                variant.unit_variant()?;
+                Ok(TypeTag::Bool)
+            }
+            DeTypeTag::U8 => {
+                variant.unit_variant()?;
+                Ok(TypeTag::U8)
+            }
+            DeTypeTag::U64 => {
+                variant.unit_variant()?;
+                Ok(TypeTag::U64)
+            }
+            DeTypeTag::U128 => {
+                variant.unit_variant()?;
+                Ok(TypeTag::U128)
+            }
+            DeTypeTag::Address => {
+                variant.unit_variant()?;
+                Ok(TypeTag::Address)
+            }
+            DeTypeTag::Signer => {
+                variant.unit_variant()?;
+                Ok(TypeTag::Signer)
+            }
+            DeTypeTag::Vector => Ok(TypeTag::Vector(Box::new(
+                variant.newtype_variant_seed(NestedTypeTag(self.0))?,
+            ))),
+            DeTypeTag::Struct => Ok(TypeTag::Struct(
+                variant.newtype_variant_seed(NestedStructTag(self.0))?,
+            )),
+        }
+    }
+}
+
+struct NestedTypeTag(u8);
+
+impl<'de> DeserializeSeed<'de> for NestedTypeTag {
+    type Value = TypeTag;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        println!("tt {}", self.0);
+        if self.0 >= MAX_DE_TYPE_TAG_NESTING {
+            return Err(de::Error::custom(format!(
+                "Hit TypeTag nesting limit during deserialization: {}",
```

Comment:
Similarly, I would change the phrasing here.  `Hit` isn't necessarily a very portable wording.

`"TypeTag has more than {} nested types during deserialization: {}"`